source/2.4/man/nonlinear/int3d.man
.TH int3d G "April 1993" "Scilab Group" "Scilab Function"
.so ../sci.an
.SH NAME
int3d - definite 3D integral by quadrature and cubature method
.SH CALLING SEQUENCE
.nf
[result,err]=int3d(X,Y,Z,f [,nf[,params]])
.fi
.SH PARAMETERS
.TP 10
X 
: a 4 by \fVNUMTET\fR array containing the abscissae of the vertices
  of the \fVNUMTET\fR tetrahedrons.
.TP 10
Y 
: a 4 by \fVNUMTET\fR array containing the ordinates of the vertices
  of the \fVNUMTET\fR tetrahedrons.
.TP
Z 
: a 4 by \fVNUMTET\fR array containing the third coordinates of the vertices
  of the \fVNUMTET\fR tetrahedrons.
.TP
f
: external (function or list or string) defining the integrand
\fVf(xyz,nf)\fR, where \fVxyz\fR is the vector of a point coordinates
and nf the number functions
.TP
nf
: the number of function to integate (default is 1)
.TP
params
: real vector \fV[minpts, maxpts, epsabs, epsrel]\fR. default
value is \fV[0, 1000, 0.0, 1.d-5]\fR.
.RS
.TP
epsabs
: Desired bound on the absolute error.  
.TP
epsrel
: Desired bound on the relative error.  
.TP
minpts
: Minimum number of function evaluations.
.TP
maxpts 
: Maximum number of function evaluations.  The number of
function evaluations over each subregion is 43
.RE
.TP
result
: the integral value,or vector of the integral values.
.TP
err
: Estimates of absolute errors.

.SH DESCRIPTION

The function calculates an approximation to a given
vector of definite integrals
.nf
I  I  I (F ,F ,...,F )      dx(3)dx(2)dx(1),
          1  2      numfun
.fi
where the region of integration is a collection of
NUMTET tetrahedrons and where 
.nf
F = F (X(1),X(2),X(3)), J = 1,2,...,NUMFUN.
 J   J
.fi

A globally adaptive strategy is applied in order to
compute approximations \fVresult(k)\fR
hopefully satisfying, for each component of I, the
following claim for accuracy:
\fVABS(I(K)-RESULT(K))<=MAX(EPSABS,EPSREL*ABS(I(K)))\fR


\fVint3d\fR repeatedly subdivides the tetrahedrons with greatest
estimated errors and estimates the integrals and the errors over the
new subtetrahedrons until the error request is met or \fVMAXPTS\fR
function evaluations have been used.

A 43 point integration rule with all evaluation points inside the
tetrahedron is applied. The rule has polynomial degree 8.

If the values of the input parameters \fVEPSABS\fR or \fVEPSREL\fR are
selected great enough, an integration rule is applied over each
tetrahedron and the results are added up to give the approximations
\fVRESULT(K)\fR. No further subdivision of the tetrahedrons will then
be applied.

When \fVint3d\fR computes estimates to a vector of
integrals, all components of the vector are given
the same treatment. That is, \fVI(Fj)\fR and \fVI(Fk)\fR for

\fVj\fR not equal to \fVk\fR, are estimated with the same
subdivision of the region of integration.
For integrals with enough similarity, we may save
time by applying \fVint3d\fR to all integrands in one call.
For integrals that varies continuously as functions of
some parameter, the estimates produced by \fVint3d\fR  will
also vary continuously when the same subdivision is
applied to all components. This will generally not be
the case when the different components are given
separate treatment.

On the other hand this feature should be used with
caution when the different components of the integrals
require clearly different subdivisions.

.SH EXAMPLES
.nf
X=[0;1;0;0];
Y=[0;0;1;0];
Z=[0;0;0;1];
deff('v=f(xyz,numfun)','v=exp(xyz''*xyz)')
[RESULT,ERROR]=int3d(X,Y,Z,'int3dex')
// computes the integrand exp(x*x+y*y+z*z) over the 
//tetrahedron (0.,0.,0.),(1.,0.,0.),(0.,1.,0.),(0.,0.,1.)
.fi
.SH SEE ALSO
intc, intl, int3d
.SH REFERENCES
Fortran routine dqtet.f
.LP
Authors: 
.nf
Jarle Berntsen, The Computing Centre,
University of Bergen, Thormohlens gt. 55,
N-5008 Bergen, Norway
Phone..  47-5-544055
Email..  jarle@eik.ii.uib.no

Ronald Cools, Dept. of Computer Science,
Katholieke Universiteit Leuven, Celestijnenlaan 200A,
B-3030 Heverlee, Belgium
Phone..  32-16-201015 (3562)
Email..  ronald@cs.kuleuven.ac.be

Terje O. Espelid, Department of Informatics,
University of Bergen, Thormohlens gt. 55,
N-5008 Bergen, Norway
Phone..  47-5-544180
Email..  terje@eik.ii.uib.no
.fi


source/2.4/routines/interf/int3d.f
      subroutine int3d
C     --------------------------------------------
c     Scilab int3d 
c      implicit undefined (a-z)
c     Copyright INRIA
      include '../stack.h'
      character*(5) fname
      character*(nlgh+1) namef
      integer iero 
      common/ierajf/iero
      external bint3d,fint3d
      double precision epsabs,epsrel,result,abserr
      logical getexternal,getrmat,type ,cremat,checkval,getscalar
      integer topk,lr,katop,kydot,top2,lra,lrb,lc,gettype,top0
c
      integer iadr,sadr
      external setfint3d
c
      iadr(l)=l+l-1
      sadr(l)=(l/2)+1
      fname='int3d'
      if(rhs.lt.4) then
         call error(39)
         return
      endif
      type=.false.
      top0=top-rhs+1
      topk=top
      if(rhs.eq.6) then
c      [minpts,maxpts,epsabs,epsrel]
         if (.not.getrmat(fname,topk,top,m1,n1,lr))  return
         if(.not.checkval(fname,m1*n1,4)) return
         minpts=max(int(stk(lr)),0)
         maxpts=max(int(stk(lr+1)),0)
         epsabs=max(stk(lr+2),0.0d0)
         epsrel=max(stk(lr+3),0.0d0)
         if(maxpts.lt.minpts) then
            err= 5
            call error(36)
         endif
         top=top-1
      else
         minpts=0
         maxpts=1000
         epsabs=0.0d0
         epsrel=1.D-5
      endif
c
      if(gettype(top).eq.1) then
         if(.not.getscalar(fname,topk,top,lnum)) return
         numfun=stk(lnum)
         top=top-1
      else
         numfun=1
      endif
      top2=top
      if (.not.getexternal(fname,topk,top,namef,type,setfint3d)) return
      kext=top
      top=top-1
      
      if (.not.getrmat(fname,topk,top,mz,nz,lz))  return
      if(.not.checkval(fname,mz,4)) return

      top=top-1
      if (.not.getrmat(fname,topk,top,my,ny,ly))  return
      if(.not.checkval(fname,my,4)) return
      top=top-1
      if (.not.getrmat(fname,topk,top,mx,nx,lx))  return
      if(.not.checkval(fname,mx,4)) return

      if(.not.checkval(fname,nx,ny)) return
      if(.not.checkval(fname,nx,nz)) return
      numtet=nx

      maxpts=max(maxpts,43*numtet)

      top=top2
c     form matrix ver
      maxsub = 7*(maxpts-43*numtet)/(8*43) + numtet
      lenver=maxsub
      top=top+1
      if (.not.cremat(fname,top,0,3,4*lenver,lver,lc)) return
      call dcopy(4*numtet,stk(lx),1,stk(lver),3)
      call dcopy(4*numtet,stk(ly),1,stk(lver+1),3)
      call dcopy(4*numtet,stk(lz),1,stk(lver+2),3)

c     allocate matrix for result and abserr
      top=top+1
      if (.not.cremat(fname,top,0,numfun,2,lres,lc)) return

c     form arg shape for external
      top=top+1
      kxyz=top
      if (.not.cremat(fname,top,0,3,1,lxyz,lc)) return

c     tableaux de travail 
      mdiv=1
      nw=maxsub*(2*numfun+1) + 7*max(8*mdiv,numtet)*numfun + 1
      top=top+1
      if (.not.cremat(fname,top,0,1,nw,lwork,lc)) return

c     tableau de travail entier necessaire 
      niw=lenver + mdiv
      top=top+1
      if (.not.cremat(fname,top,0,1,iadr(niw)+1,liwork,lc)) return

c
c     external scilab
c
      top=top+1
      ipal=iadr(lstk(top))
      istk(ipal)=1
      istk(ipal+1)=ipal+2
      istk(ipal+2)=kext
      istk(ipal+3)=kxyz
      lstk(top+1)=sadr(ipal+4)

      irestar=0
      if(type) then
         call dcutet(fint3d,numfun,stk(lver),numtet,minpts,maxpts,epsabs
     $        ,epsrel,lenver,nw,irestar,stk(lres),stk(lres+numfun),neval
     $        ,ifail,stk(lwork),stk(liwork))
      else
         call dcutet(bint3d,numfun,stk(lver),numtet,minpts,maxpts,epsabs
     $        ,epsrel,lenver,nw,irestar,stk(lres),stk(lres+numfun),neval
     $        ,ifail,stk(lwork),stk(liwork))
      endif
      if(err.gt.0)return
      if(ifail.gt.0) then
         if(ifail.eq.1) then
c     MAXPTS was too small to obtain the required accuracy
            call msgs(85,0)
         elseif(ifail.eq.3) then
c     volume of one of the  given tetrahedrons is zero
            buf='volume of one of the  given tetrahedrons is zero'
            call error(999)
         else
            buf='inexpected error: please report'
            call error(999)
         endif
      endif
      top=top0
      if (.not.cremat(fname,top,0,numfun,1,lr1,lc)) return
      call dcopy(numfun,stk(lres),1,stk(lr1),1)

      if(lhs.ge.2) then
         top=top+1
         if (.not.cremat(fname,top,0,numfun,1,lr2,lc)) return
         call dcopy(numfun,stk(lres+numfun),1,stk(lr2),1)
      endif
      if(lhs.eq.3) then
         top=top+1
         if (.not.cremat(fname,top,0,1,1,lr2,lc)) return
         stk(lr2)=neval
      endif
      return
      end

